controllers/grafananotificationtemplate_controller.go

```diff
@@ -0,0 +1,260 @@
+			Time: time.Now(),
+		},
+	}
+
+	if len(applyErrors) == 0 {
+		condition.Status = metav1.ConditionTrue
+		condition.Reason = conditionApplySuccessful
+		condition.Message = fmt.Sprintf("Notification template was successfully applied to %d instances", len(instances))
+	} else {
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = conditionApplyFailed
+
+		var sb strings.Builder
+		for i, err := range applyErrors {
+			sb.WriteString(fmt.Sprintf("\n- %s: %s", i, err))
+		}
+
+		condition.Message = fmt.Sprintf("Notification template failed to be applied for %d out of %d instances. Errors:%s", len(applyErrors), len(instances), sb.String())
+	}
+	meta.SetStatusCondition(&notificationTemplate.Status.Conditions, condition)
+	if notificationTemplate.ResyncPeriodHasElapsed() {
+		notificationTemplate.Status.LastResync = metav1.Time{Time: time.Now()}
+	}
+
+	return ctrl.Result{RequeueAfter: notificationTemplate.Spec.ResyncPeriod.Duration}, nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) reconcileWithInstance(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return fmt.Errorf("building grafana client: %w", err)
+	}
+
+	trueRef := "true" //nolint:goconst
+	editable := true
+	if notificationTemplate.Spec.Editable != nil && !*notificationTemplate.Spec.Editable {
+		editable = false
+	}
+
+	var updatedNT models.NotificationTemplateContent
+	updatedNT.Template = notificationTemplate.Spec.Template
+	params := provisioning.NewPutTemplateParams().WithName(notificationTemplate.Spec.Name).WithBody(&updatedNT)
+	if editable {
+		params.SetXDisableProvenance(&trueRef)
+	}
+	_, err = cl.Provisioning.PutTemplate(params) //nolint:errcheck
+	if err != nil {
+		return fmt.Errorf("creating or updating notification template: %w", err)
+	}
+	return nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) getNotificationTemplateByName(ctx context.Context, instance *grafanav1beta1.Grafana, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) (models.NotificationTemplate, error) {
+	cl, err := client2.NewGeneratedGrafanaClient(ctx, r.Client, instance)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("building grafana client: %w", err)
+	}
+
+	params := provisioning.NewGetTemplateParams().WithName(notificationTemplate.Spec.Name)
+	remote, err := cl.Provisioning.GetTemplateWithParams(params)
+	if err != nil {
+		return models.NotificationTemplate{}, fmt.Errorf("getting notification template: %w", err)
+	}
+	return *remote.Payload, nil
+}
+
+func (r *GrafanaNotificationTemplateReconciler) finalize(ctx context.Context, notificationTemplate *grafanav1beta1.GrafanaNotificationTemplate) error {
+	r.Log.Info("Finalizing GrafanaNotificationTemplate")
+
+	instances, err := r.GetMatchingInstances(ctx, notificationTemplate, r.Client)
```

Comment:
This differs from `GetScopedMatchingInstances` by performing a ready check before returning the list of instances, and ignoring namespace boundaries.
```suggestion
	instances, err := r.GetAllMatchingInstances(ctx, r.Client, notificationTemplate)
```